core/field_number.js

```diff
@@ -51,316 +45,267 @@ goog.require('goog.userAgent');
+    var spliceValue = e.target.innerText;
+    // Old value of the text field
+    var oldValue = this.htmlInput_.value;
+    // Determine the selected portion of the text field
+    var selectionStart = this.htmlInput_.selectionStart;
+    var selectionEnd = this.htmlInput_.selectionEnd;
+
+    // Splice in the new value
+    var newValue = oldValue.slice(0, selectionStart) + spliceValue +
+        oldValue.slice(selectionEnd);
+
+    // Set new value and advance the cursor
+    this.updateDisplay_(newValue, selectionStart + spliceValue.length);
+
+    // This is just a click.
+    Blockly.Touch.clearTouchIdentifier();
+
+    // Prevent default to not lose input focus
+    e.preventDefault();
+  }
 
-  // This is just a click.
-  Blockly.Touch.clearTouchIdentifier();
+  /**
+   * Call for when the num-pad erase button is touched.
+   * Determine what the user is asking to erase, and erase it.
+   * @param {Event} e DOM event triggering the touch.
+   */
+  numPadEraseButtonTouch(e) {
+    // Old value of the text field
+    var oldValue = this.htmlInput_.value;
+    // Determine what is selected to erase (if anything)
+    var selectionStart = this.htmlInput_.selectionStart;
+    var selectionEnd = this.htmlInput_.selectionEnd;
+
+    // If selection is zero-length, shift start to the left 1 character
+    if (selectionStart == selectionEnd) {
+      selectionStart = Math.max(0, selectionStart - 1);
+    }
 
-  // Prevent default to not lose input focus
-  e.preventDefault();
-};
+    // Cut out selected range
+    var newValue = oldValue.slice(0, selectionStart) +
+        oldValue.slice(selectionEnd);
 
-/**
- * Call for when the num-pad erase button is touched.
- * Determine what the user is asking to erase, and erase it.
- * @param {Event} e DOM event triggering the touch.
- */
-Blockly.FieldNumber.numPadEraseButtonTouch = function(e) {
-  // Old value of the text field
-  var oldValue = Blockly.FieldTextInput.htmlInput_.value;
-  // Determine what is selected to erase (if anything)
-  var selectionStart = Blockly.FieldTextInput.htmlInput_.selectionStart;
-  var selectionEnd = Blockly.FieldTextInput.htmlInput_.selectionEnd;
-
-  // If selection is zero-length, shift start to the left 1 character
-  if (selectionStart == selectionEnd) {
-    selectionStart = Math.max(0, selectionStart - 1);
-  }
+    this.updateDisplay_(newValue, selectionStart);
 
-  // Cut out selected range
-  var newValue = oldValue.slice(0, selectionStart) +
-      oldValue.slice(selectionEnd);
+    // This is just a click.
+    Blockly.Touch.clearTouchIdentifier();
 
-  Blockly.FieldNumber.updateDisplay_(newValue, selectionStart);
+    // Prevent default to not lose input focus which resets cursors in Chrome
+    e.preventDefault();
+  }
 
-  // This is just a click.
-  Blockly.Touch.clearTouchIdentifier();
+  doClassValidation_(
+    newValue
+  ) {
+    if (newValue === '') return newValue;
+    return super.doClassValidation_(newValue);
```

Comment:
I think the validation for the scratch numbers is actually different from core. E.g. in core we allow folks to type in letters and we just highlight the field in red. But in scratch they don't allow you to type them at all.

I think you might be able to grab the logic from `getNumRestrictor`.

Reply:
Done, but in `onHtmlInputKeyDown_` to duplicate Scratch's behavior of not allowing bad input at all.